server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -178,19 +177,28 @@ private void executeKnnVectorQuery(SearchContext context) throws IOException {
         List<DfsKnnResults> knnResults = new ArrayList<>(knnVectorQueryBuilders.size());
         for (int i = 0; i < knnSearch.size(); i++) {
             Query knnQuery = searchExecutionContext.toQuery(knnVectorQueryBuilders.get(i)).query();
-            TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
-            CollectorManager<Collector, Void> collectorManager = new SingleThreadCollectorManager(topScoreDocCollector);
-            if (context.getProfilers() != null) {
+            if (context.getProfilers() == null) {
+                // without profiling enabled we use a CollectorManager that supports parallelized execution
+                CollectorManager<TopScoreDocCollector, TopDocs> collectorManager = TopScoreDocCollector.createSharedManager(
+                    knnSearch.get(i).k(),
+                    null,
+                    Integer.MAX_VALUE
+                );
+                TopDocs topDocs = context.searcher().search(knnQuery, collectorManager);
+                knnResults.add(new DfsKnnResults(topDocs.scoreDocs));
+            } else {
+                TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
```

Comment:
Absolutely agree. I tried that too here, but without making changes to the InternalProfileCollectorManager it wasn't really clear, partly because of the missing reduce-return type there. I think in general I'd like to understand / discuss better how we think about the "reduce" method, whether it returns something like in the case of most Lucene CMs or if we want to have it return "Void" for most cases like we do now.
I can take a look at making the Profiler-Wrapper multithreaded next but I had a feeling we can postpone that.

Reply:
I am bumping into the same issues with #96622 . I think DFSPhase can be an exception because of the simplicity of the collectors being used in it, when it comes to the return type. It becomes much more complicated in the `QueryPhase` and given that the profiler is shared across the two phases, we need to find a good way to address this that works on both sides. In general, the current thinking is that when you have a complex hierarchy of collectors, the return type and the generics are mostly on the way and we should not rely on it.